### PR TITLE
resolve: Support for DNS-over-TLS

### LINKED ISSUE #5671: Please add support for DNS-over-TLS to systemd-resolved
### Submission type

  - [ ] Bug report
  - [x] Request for enhancement (RFE)

### Details

Please add support to systemd-resolved for talking to servers that support [RFC 7858](https://tools.ietf.org/html/rfc7858) DNS-over-TLS on port 853.

Publicly available :853 test servers are listed on the [sinodun wiki](https://portal.sinodun.com/wiki/display/TDNS/DNS-over-TLS+test+servers).

### PR DESCRIPTION
**Pull request by irtimmer**
This adds support for DNS-over-TLS (RFC 7858), which can be enabled by setting _DNSTLS_ to _yes_ or _allow_downgrade_ (although due the timeout of 10s downgrading can be slow) in _resolved.conf_, and long lived TCP connections.

Instead of reopening a new TCP connection for every DNS request using TCP, it will first check for already opened stream to the same server. A list of all transactions per stream is kept in a list beside a list containing all DNS packets requested to be written by the transactions. Also a reference to the stream is kept to keep it open until it's closed by the server or it's timed out. Maybe this could later be extended with edns-tcp-keepalive (RFC 7828), but I couldn't quickly find a server supporting this extension. Also connection errors are counted once per stream instead of per transaction.

To support different protocols over which DNS can be provide, the feature level is split into a feature level and protocol level. This makes support for EDNS, DNSSEC independent from the transport protocol (TLS, UDP, TCP).

For supporting TLS this uses GNUTLS which is already used by journald. Currently it won't validate certificates, connect only over port 583 and can only be configured global and not per link. To keep reconnections over TLS fast, it use TCP Fast Open (in a separate commit), TLS False Start and TLS Session Resumption. Although a quick test showed not everything is supported by the current public DNS-over-TLS severs.

For TCP Fast Open instead of using _connect()_ the socket address is passed to the DnsStream, while _sendmsg()_ is used to connect and send data at the same time.

Closes #5671 and #7617

### CODE REVIEW

File: src/resolve/resolved-dns-packet.h [hunk 1]
```diff
@@ -81,6 +81,9 @@ struct DnsPacket {
         /* For support of truncated packets */
         DnsPacket *more;
 
+        /* For support of packets in queue */
+        LIST_FIELDS(DnsPacket, packets_in_queue);
+
```
Comment (poettering) [inline src/resolve/resolved-dns-packet.h]: hmm, I wonder if this is a safe thing to do. So far we considered packets more or less "immutable" after they have been created once, thus encouraging that consumers just add refs if they want to keep them around and don't have to be afraid that they change under their accesses too wildly. Coding using them can do with the packets whatever it wants really, in the current code and in future code and as long as they take a ref they know that the packet is gonna stay pinned and remain ina good state.

However, this property is somewhat affected by this embedded chain, as this means the packets can be used in a specific context only once: you can enqueue it in exactly one DnsStream queue exactly once. If any other code gets ahold of the DnsPacket, then it can't use it in that context really, and if it does by some accident weird things will happen…

Now, in the current code I can't see why one ever would want to enqueue the same DnsPacket in more than one stream, or more than once in the same stream, but I still would prefer if we'd not embedd the linked list structure here, just to keep the code more versatile and generic for future changes.

Hence, instead of embedding a linked list here, I think it would be nice to just use an `OrderedSet` object for maintaining the queue of packets to write. I figure the code is probably going to be shorter too that way, and a lot more generic, as enqueing the same DnsPacket objects into multiple streams or into the same stream more than once would be fully safe then.

File: src/resolve/resolved-dns-transaction.c [hunk 2]
```diff
@@ -509,8 +508,68 @@ static int on_stream_complete(DnsStream *s, int error) {
...
+        /* Ignore incorrect transaction id as transaction can have been canceled */
+        if (ct)
+                r = on_dns_stream_packet(t, p);
+        else {
+                if (dns_packet_validate_reply(p) <= 0) {
+                        log_debug("Invalid TCP reply packet.");
+                        on_stream_complete(s, 0);
+                        return 0;
+                }
+                return 0;
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: you can dorp the inner `return 0`...

File: src/resolve/resolved-dns-transaction.c [hunk 3]
```diff
@@ -509,8 +508,68 @@ static int on_stream_complete(DnsStream *s, int error) {
...
+        p = s->read_packet;
+        s->read_packet = NULL;
+        s->n_read = 0;
+
+        /* Find corresponding transaction for this answer */
+        LIST_FOREACH(transactions_by_stream, t, s->transactions)
+                if (DNS_PACKET_ID(p) == t->id) {
+                        ct = t;
+                        break;
+                }
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: you can use `hashmap_get(s->manager->dns_transactions, UINT_TO_PTR(DNS_PACKET_ID(p))` to get the transaction for the specified ID directly in O(1), no need to search linearly here

File: src/resolve/resolved-dns-transaction.c [hunk 4]
```diff
@@ -563,28 +625,37 @@ static int dns_transaction_open_tcp(DnsTransaction *t) {
...
-                return r;
-        fd = -1;
+                r = dns_stream_new(t->scope->manager, &s, t->scope->protocol, fd);
+                if (r < 0)
+                        return r;
+
+                fd = -1;
+
+                if (t->server)
+                        s->server = dns_server_ref(t->server);
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: dns_server_ref() returns NULL if you pass in NULL (most our _ref() calls do), hence you can drop the if check here, it's redundant

File: src/resolve/resolved-manager.c [hunk 5]
```diff
@@ -1224,6 +1224,19 @@ DnsScope* manager_find_scope(Manager *m, DnsPacket *p) {
...
+        DnsStream *i;
+
+        assert(m);
+        assert(s);
+
+        LIST_FOREACH(streams, i, m->dns_streams)
+                if (i->server == s)
+                        return i;
+
+        return NULL;
```
Comment (poettering) [inline src/resolve/resolved-manager.c]: i think it would make a ton of sense to embedd a ref to a DnsStream in the DnsServer structure (or even a `Set` object with multiple of them). Since we want to share connections now, it only makes sense if we maintain a per-server stream if we can as part of the server structure. I figure in the long run it might make sense to keep more than one stream per server open, hence maybe make this a Set* and allocate multiple streams per service.

I figure a simple algorithm of "when we want to send a new DnsPacket, look for suitable existing DnsStream objects to the DnsServer, and use the first one with an empty queue. If none has an empty queue allocate a new DnsStream object until a max of 5 streams per server is reached, in which case enqueue into the stream with the shortest queue" or so might be simple enough to implement for now and good for the beginning

If the ref to the DnsStream is embedded in the DnsServer we don't have to search the full list of streams here...

File: src/resolve/resolved-dns-scope.c [hunk 6]
```diff
@@ -306,55 +306,60 @@ static int dns_scope_socket(
                 int family,
                 const union in_addr_union *address,
                 DnsServer *server,
-                uint16_t port) {
+                uint16_t port,
+                union sockaddr_union **socket_address) {
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c]: for new code we try to follow the rule that output parameters are prefixed with `ret_`, i.e. please rename `socket_address` to `ret_socket_address` here

File: src/resolve/resolved-dns-scope.c [hunk 7]
```diff
@@ -381,30 +386,35 @@ static int dns_scope_socket(
...
                                 return -errno;
                 }
         }
 
-        r = connect(fd, &sa.sa, salen);
-        if (r < 0 && errno != EINPROGRESS)
-                return -errno;
+        if (socket_address) {
+                *socket_address = sa;
+                sa = NULL;
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c]:
```c
*socket_address = TAKE_PTR(sa);
```

File: src/resolve/resolved-dns-scope.c [hunk 8]
```diff
@@ -306,55 +306,60 @@ static int dns_scope_socket(
...
         _cleanup_close_ int fd = -1;
-        union sockaddr_union sa = {};
+        _cleanup_free_ union sockaddr_union *sa;
         socklen_t salen;
         static const int one = 1;
         int r, ifindex;
 
         assert(s);
 
+        sa = new0(union sockaddr_union, 1);
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c]: hmm, why do this with dynamic memory btw? I think the the callers should simply pass in a pre-allocated buffer, which this function fills on on success...

It's fine of we copy the structure around a bit... Today's CPUs excel at copying stuff around, and this structure isn't < 200 bytes iirc

File: src/resolve/resolved-dns-stream.c [hunk 9]
```diff
@@ -183,6 +183,36 @@ static int dns_stream_identify(DnsStream *s) {
...
+        assert(s);
+        assert(iov);
+
+        if (s->tfo_salen) {
+                struct msghdr hdr = {};
+
+                hdr.msg_iov = iov;
+                hdr.msg_iovlen = iovcnt;
+                hdr.msg_name = &s->tfo_address->sa;
+                hdr.msg_namelen = s->tfo_salen;
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: please, fold this into the structure initialization above. It's not just nicer to look at but compilers usually generate better code too that way. i.e.:

```c
struct msghdr hdr = {
        .msg_iov = iov,
        .msg_iovlen …
…
}
```

File: src/resolve/resolved-dns-stream.c [hunk 10]
```diff
@@ -183,6 +183,36 @@ static int dns_stream_identify(DnsStream *s) {
...
+                r = sendmsg(s->fd, &hdr, MSG_FASTOPEN);
+                if (r < 0) {
+                        if (errno == EOPNOTSUPP) {
+                                mfree(s->tfo_address);
+                                s->tfo_salen = 0;
+                                errno = EAGAIN;
+                                connect(s->fd, (struct sockaddr*) &s->tfo_address->sa, s->tfo_salen);
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: hmm, no error checking?

File: src/resolve/resolved-dns-stream.c [hunk 11]
```diff
@@ -183,6 +183,36 @@ static int dns_stream_identify(DnsStream *s) {
...
+                struct msghdr hdr = {};
+
+                hdr.msg_iov = iov;
+                hdr.msg_iovlen = iovcnt;
+                hdr.msg_name = &s->tfo_address->sa;
+                hdr.msg_namelen = s->tfo_salen;
+                r = sendmsg(s->fd, &hdr, MSG_FASTOPEN);
+                if (r < 0) {
+                        if (errno == EOPNOTSUPP) {
+                                mfree(s->tfo_address);
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: mfree() is a wrapper around free() with the single purpose of also returning NULL, so that you can write stuff like:

```c
p = mfree(p);
```

but, if you don't assign the result to anything, then there's no point in using it, you might as well use regular free()

File: src/resolve/resolved-dns-stream.c (hunk 10 above)
Comment (poettering) [reply src/resolve/resolved-dns-stream.c]: also, and more importantly, you pass a ptr to an address you just freed a few lines up...

Also, the cast should not be necessary, as the `->sa` field already should have the right type

File: src/resolve/resolved-dns-stream.c [hunk 12]
```diff
@@ -183,6 +183,36 @@ static int dns_stream_identify(DnsStream *s) {
         return 0;
 }
 
+static ssize_t dns_stream_writev(DnsStream *s, struct iovec *iov, int iovcnt) {
+        ssize_t r;
+
+        assert(s);
+        assert(iov);
+
+        if (s->tfo_salen) {
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: we usually avoid relying on C's downgrade-to-bool feature for numeric values that aren't actually conceptually booleans (except for pointer checks). Since this is a byte length, hence please use "if (s->tfo_salen > 0)"

File: src/resolve/resolved-dns-stream.c [hunk 13]
```diff
@@ -183,6 +183,36 @@ static int dns_stream_identify(DnsStream *s) {
         return 0;
 }
 
+static ssize_t dns_stream_writev(DnsStream *s, struct iovec *iov, int iovcnt) {
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: iov can be const, no?

It's a bit weird that writev()'s iovcnt parameter is an "int" btw... it really should be a size_t, since it counts structures in memory. If we define a wrapper like this one we should really fix that, and hide the "int" thing internally

File: src/resolve/resolved-dns-stream.c [hunk 14]
```diff
@@ -197,9 +227,12 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
 
         assert(s);
 
-        r = dns_stream_identify(s);
-        if (r < 0)
-                return dns_stream_complete(s, -r);
+        /* only identify after connecting */
+        if (!s->tfo_salen) {
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: as above, please use `if (s->tfo_salen == 0)` or so

File: src/resolve/resolved-dns-stream.c [hunk 15]
```diff
@@ -341,6 +374,9 @@ DnsStream *dns_stream_unref(DnsStream *s) {
                 s->manager->n_dns_streams--;
         }
 
+        if (s->tfo_address)
+                mfree(s->tfo_address);
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: no mfree() here, as you aren't interested in the return value. Moreover, the if check is redundant, both glibc free() and our mfree() wrapper are fine with a NULL parameter, and become NOPs in that case

File: src/resolve/resolved-dns-server.c [hunk 16]
```diff
@@ -213,14 +213,15 @@ void dns_server_move_back_and_unmark(DnsServer *s) {
         }
 }
```
Comment (poettering) [inline src/resolve/resolved-dns-server.c]: hmm, can you elaborate on the reasoning of this commit? I mean, I'd assume that servers that can do DNS-over-TLS can also do EDNS0+DO+LARGE just fine, hence I always assumed we'd add DNS-over-TLS simply as another feature level on top of all the others?

Can you explain what made you split this up?

Note that the feature level the way it was currently implemented implemented a state machine of sorts, and not just a linear level. Specifically, we'd start out at LARGE, and then when encountering too many failures we'd go to DO, then to EDNS0, then to UDP, then to TCP, and then back to UDP and then TCP again and so on. Or in other words:

```
   LARGE → DO → EDNS0 → UDP → TCP
                         ↑     ↓     
                            ←
```

And I'd always have assumed it would now become:

```
   TLS → LARGE → DO → EDNS0 → UDP → TCP
                               ↑     ↓     
                                  ←
```

Where TLS is supposed to mean TLS-over-TCP, and LARGE, DO, EDN0 all mean extension of plain UDP.

File: meson.build [hunk 17]
```diff
@@ -2801,6 +2814,7 @@ status = [
         'symbolic gateway hostnames:        @0@'.format(', '.join(gateway_hostnames)),
 
         'default DNSSEC mode:               @0@'.format(default_dnssec),
+        'default DNS-over-TLS mode:         @0@'.format(default_dnstls),
```
Comment (poettering) [inline meson.build]: i figure using gnutls for this is fine for now. I have the suspicion now that everybody appears to settle on OpenSSL these days, and we should gradually move over to that.

**Comment by poettering** [comment]
> Although I still think it's a good idea to split it up as I don't think expecting complete DNSSEC support when using DNS-over-TLS is a valid assumption. RFC 7858 even explicitly state that these two are independent.

Hmm, I see, makes sense. But I'd really like to the feature level to be considered more a state machine, hence we probably should come up with a different state machine logic then for this. maybe:

```
TLS-DO → LARGE → DO → TLS-PLAIN → EDNS0 → UDP → TCP
                                           ↑     ↓     
                                              ←
```

Where TLS-DO would be the combination of TLS and DNSSEC, and TLS-PLAIN for would TLS without DNSSEC. If TLS and DNSSEC are on, we'd start with TLS-DO. If only DNSSEC is on, we'd start with LARGE, and if only TLS is on we'd start with TLS-PLAIN. If TLS is off we'd jump from DO directly to EDNS0 of course. Now, there's one major ambiguity: if both TLS and DNSSEC are in "allow-downgrade" mode, do we first give up on TLS or first give up on DNSSEC? With the above we'd first give up on TLS, and consider DNSSEC more important. This translates to preferring client-side authentication over privacy, which I think is a good thing.

File: src/resolve/resolved-dns-packet.c [hunk 18]
```diff
@@ -2333,6 +2333,36 @@ int dns_packet_is_reply_for(DnsPacket *p, const DnsResourceKey *key) {
...
+static int dns_packet_compare_func(const void *a, const void *b) {
+        const DnsPacket *x = a, *y = b;
+        int r;
+
+        if (x->size < y->size)
+                return -1;
+        if (x->size > y->size)
+                return 1;
+
+        r = memcmp(DNS_PACKET_DATA((DnsPacket*) x), DNS_PACKET_DATA((DnsPacket*) y), x->size);
```
Comment (poettering) [inline src/resolve/resolved-dns-packet.c]: you could shorten this to `return memcmp(…)` here…

File: src/resolve/resolved-dns-stream.c [hunk 19]
```diff
@@ -394,12 +418,7 @@ int dns_stream_new(Manager *m, DnsStream **ret, DnsProtocol protocol, int fd) {
...
 
-        if (s->write_packet)
-                return -EBUSY;
-
-        s->write_packet = dns_packet_ref(p);
-        s->write_size = htobe16(p->size);
-        s->n_written = 0;
-
+        dns_packet_ref(p);
+        ordered_set_put(s->write_queue, p);
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: adding an item to a set can fail. You need to check the ordered_set_put() return value hence, and only add a ref to the packet on success

File: src/resolve/resolved-dns-transaction.c [hunk 20]
```diff
@@ -486,6 +473,15 @@ static int on_stream_complete(DnsStream *s, int error) {
                 return 0;
         }
 
+        return 0;
+}
+
+static int on_dns_stream_packet(DnsTransaction *t, DnsPacket *p) {
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: mhmm, i don't like that we have two functions now named so similar: on_stream_packet() and on_dns_stream_packet. Let's clean the naming up, i.e. properly prefix the two calls with the "class" they belong to: dns_transaction_on_stream_packet() and dns_stream_on_packet()

File: src/resolve/resolved-dns-transaction.c [hunk 21]
```diff
@@ -509,8 +505,64 @@ static int on_stream_complete(DnsStream *s, int error) {
...
+                        assert_se(sd_event_now(t->scope->manager->event, clock_boottime_or_monotonic(), &usec) >= 0);
+                        dns_server_packet_lost(t->server, IPPROTO_UDP, t->current_feature_level, usec - t->start_usec);
+                }
+        }
+
+        LIST_FOREACH_SAFE(transactions_by_stream, t, n, s->transactions)
+                if (error != 0)
+                        on_transaction_stream_error(t, error);
+                else if (DNS_PACKET_ID(s->read_packet) == t->id)
+                        r = on_dns_stream_packet(t, s->read_packet);
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: hmm, we need to do something sensible with the return code here. Currently we propagate the first matching transaction, if there is one. If there can only be one, then we really should also have a `break` here, to clarify that this is the case. If there can be multiple, then we should handle possible failures on some of them in some smart way

File: src/resolve/resolved-dns-transaction.c [hunk 22]
```diff
@@ -509,8 +505,64 @@ static int on_stream_complete(DnsStream *s, int error) {
...
+}
+
+static int on_stream_packet(DnsStream *s) {
+        _cleanup_(dns_packet_unrefp) DnsPacket *p = NULL;
+        int r = 0;
+        DnsTransaction *t;
+
+        /* Take ownership of packet to be able to receive new packets */
+        p = s->read_packet;
+        s->read_packet = NULL;
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c]: we have this nice TAKE_PTR() macro now, which you can use to reduce the two lines above into:

```c
p = TAKE_PTR(s->read_packet);
```

File: src/resolve/resolved-dns-stream.c [hunk 23]
```diff
@@ -182,6 +182,38 @@ static int dns_stream_identify(DnsStream *s) {
         return 0;
 }
 
+static ssize_t dns_stream_writev(DnsStream *s, struct iovec *iov, size_t iovcnt) {
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: i understand that dns_stream_writev() is supposed to be modelled after writev(), but I'd prefer if we'd not copy it 1:1, but return negative errors as we normally do in our codebase, rather than propagate them via errno...

I figure returning negative error numbers here inside of the ssize_t would be fine, even though error numbers are typically "int"s, but let's keep this simple.

Also, the `iov` parameter really should be const, even if this means we need to cast the const away internally...

File: src/resolve/resolved-dns-stream.c [hunk 24]
```diff
@@ -366,7 +401,7 @@ DnsStream *dns_stream_ref(DnsStream *s) {
         return s;
 }
 
-int dns_stream_new(Manager *m, DnsStream **ret, DnsProtocol protocol, int fd) {
+int dns_stream_new(Manager *m, DnsStream **ret, DnsProtocol protocol, int fd, union sockaddr_union *tfo_address) {
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: the new parameter can be "const", right?

File: meson.build (hunk 17 above)
Comment (dkg) [reply meson.build]: i think using gnutls is fine here, fwiw.  knot-resolver (one of the popular dns resolvers that implements dns-over-tls) uses gnutls and has no problem with it.

**Comment by dkg** [comment]
this is great work, thanks @irtimmer !

I think it would be great to start by trying to merge the persistent TCP stream and the TCP fast-open patches so that the DNS-over-TLS patch can be evaluated independently.

some thoughts on the interface:
 * you're absolutely right to keep `DNSSEC` orthogonal to this configuration.
 * the option should be named `PrivateDNS=`, not `DNSTLS=` -- this more clearly expresses a straightforward user intent, and allows systemd to make sensible choices about what "private" means in the future.  This option should apply both to `resolved.conf` and to any `.network` stanza that allows configuration of DNS for a given link.
 * the option's values should be `none`, `opportunistic`, or `strict` -- in `opportunistic` mode, before a connection is established, it can try port 853 (TLS) in the background to see whether a session can be established -- while that attempt is happening but isn't yet resolved, it should go ahead and send queries in the clear, to avoid the delay.  if session establishment fails, it can continue to send queries in the clear.  if a TLS session is successfully established when a query comes in, then it can send queries through the TLS connection.
 * if the initial implementation doesn't do certificate checking, it shouldn't offer `strict` yet
 * i can offer a series that expands the definition of the DNS server from just an IP address (`192.0.2.4`) to an IP address and hostname (`192.0.2.4|dns.example.com`), which would enable certificate-hostname checking.  with that change, then we could enable `PrivateDNS=strict`, which would not permit queries to go out over an interface in the clear.

[RFC 7858](https://tools.ietf.org/html/rfc7858) suggests that requiring TLS connections without requiring certificate verification isn't a good idea -- if we want opportunistic crypto, that's fine and definitely better than cleartext.  but if we want to require TLS, we need to ensure we're talking to who we expect.

**Comment by poettering** [comment]
oh, please always add a comment when you force push a new version onto a PR. github sometimes doesn't notify us when this happens on its own, there's something weird there in github, so i didn't see the new version yet.

**Comment by irtimmer** [comment]
In this case I deliberately didn't comment as I'm still have to fix some of the state logic and the error codes coming from GnuTLS

**Comment by irtimmer** [comment]
I've again tried to incorporate all requested changes. I've added the TLS-DO and TLS-PLAIN in the suggested order to the server level, although the state machine logic got a little more complicated and should now look like:
```
TLS-DO  →  TLS-PLAIN
  ↓            ↓
LARGE → DO → EDNS0 → UDP → TCP
                      ↑     ↓
                         ←
```
Depending on the kind of failure (TCP/TLS error or DNS response error) it is decided whether TLS-DO is downgraded to TLS-PLAIN or LARGE. This ensures we don't retry a TLS connection if the TLS connection fails or failback to plaintext queries in case the TLS server just doesn't support DNSSEC.

I've renamed the option DNSTLS to PrivateDNS and removed the strict option. Additionally I added a commit which add support for setting the PrivateDNS option per link.

@dkg One of the main reasons I haven't yet implemented certificate checking, is because I couldn't figure out a good way to specify the hostname or SPKI pins per server. So I'm really interested in the patches to expand the DNS server definitions.

**Comment by ott** [comment]
The state machine should not be derived from queries but from separate probe queries because it's otherwise not possible to distinguish whether the query or the feature set caused the error in the response. Moreover, the feature set should be periodically reset to cope with load balancing, software upgrades, system load fluctuations and configuration changes but I think I should open a separate issue for that and we can postpone it.

**Comment by poettering** [comment]
> Moreover, the feature set should be periodically reset to cope with load balancing, software upgrades, system load fluctuations and configuration changes but I think we can postpone this until this pull request has been merged.

We already reset the feature set after a while, it's called the "grace period", and depending how stable the server is the grace period is picked to be 5min … 6h.

**Comment by shibumi** [comment]
Any update on this? Really can't wait to see this merged :)

File: src/resolve/resolved-dns-scope.c [hunk 25]
```diff
@@ -306,10 +306,11 @@ static int dns_scope_socket(
                 int family,
                 const union in_addr_union *address,
                 DnsServer *server,
-                uint16_t port) {
+                uint16_t port,
+                union sockaddr_union *ret_socket_address) {
```
Comment (poettering) [inline src/resolve/resolved-dns-scope.c:316]: a quick comment somewhere which explains this "delayed connect logic" a bit, i.e. that the caller needs to do the connect on its own later on (and why he would choose to do so) would be nice.

File: src/resolve/resolved-dns-stream.c [hunk 26]
```diff
@@ -386,6 +422,7 @@ int dns_stream_new(Manager *m, DnsStream **ret, DnsProtocol protocol, int fd) {
 
         s->n_ref = 1;
         s->fd = -1;
+        s->tfo_salen = 0;
```
Comment (poettering) [inline src/resolve/resolved-dns-stream.c]: no need to initialize this, as we use new0() above, which initializes this anyway. The only fields that need explicit initialization are the ones those different from zero

File: src/network/networkd-network-gperf.gperf [hunk 27]
```diff
@@ -58,6 +58,7 @@ Network.Domains,                        config_parse_domains,
 Network.DNS,                            config_parse_dns,                               0,                             0
 Network.LLMNR,                          config_parse_resolve_support,                   0,                             offsetof(Network, llmnr)
 Network.MulticastDNS,                   config_parse_resolve_support,                   0,                             offsetof(Network, mdns)
+Network.PrivateDNS,                     config_parse_private_dns_mode,                  0,                             offsetof(Network, private_dns_mode)
```
Comment (poettering) [inline src/network/networkd-network-gperf.gperf:63]: this needs some brief mention in the man page

File: src/resolve/resolvectl.c [hunk 28]
```diff
@@ -2337,6 +2409,7 @@ static void native_help(void) {
                "  domain [LINK [DOMAIN...]]    Get/set per-interface search domain\n"
                "  llmnr [LINK [MODE]]          Get/set per-interface LLMNR mode\n"
                "  mdns [LINK [MODE]]           Get/set per-interface MulticastDNS mode\n"
+               "  privatedns [LINK [MODE]]     Get/set per-interface PrivateDNS mode\n"
```
Comment (poettering) [inline src/resolve/resolvectl.c:2401]: also needs some brief mention in the man page

File: src/resolve/resolved-gperf.gperf [hunk 29]
```diff
@@ -23,5 +23,6 @@ Resolve.Domains,         config_parse_search_domains,         0,
 Resolve.LLMNR,           config_parse_resolve_support,        0,                   offsetof(Manager, llmnr_support)
 Resolve.MulticastDNS,    config_parse_resolve_support,        0,                   offsetof(Manager, mdns_support)
 Resolve.DNSSEC,          config_parse_dnssec_mode,            0,                   offsetof(Manager, dnssec_mode)
+Resolve.PrivateDNS,      config_parse_private_dns_mode,       0,                   offsetof(Manager, private_dns_mode)
```
Comment (poettering) [inline src/resolve/resolved-gperf.gperf:26]: needs a brief description in the man page

File: src/resolve/resolved-dns-scope.c (hunk 25 above)
Comment (irtimmer) [reply src/resolve/resolved-dns-scope.c:316]: a comment is added to the non-static method dns_scope_socket_tcp

**Comment by irtimmer** [comment]
I've fixed the small issues and made an attempt in updating the man pages. But suggestions to improve spelling or grammar are more than welcome

File: man/resolvectl.xml [hunk 30]
```diff
@@ -282,8 +283,8 @@
         <listitem><para>Revert the per-interface DNS configuration. If the DNS configuration is reverted all
         per-interface DNS setting are reset to their defaults, undoing all effects of <option>dns</option>,
         <option>domain</option>, <option>llmnr</option>, <option>mdns</option>, <option>dnssec</option>,
-        <option>nta=</option>. Note that when a network interface disappears all configuration is lost automatically,
-        an explicit reverting is not necessary in that case.</para></listitem>
+        <option>privatends</option>, <option>nta=</option>. Note that when a network interface disappears all
```
Comment (poettering) [inline man/resolvectl.xml]: typo: privatends → privatedns

**Comment by poettering** [comment]
found one typo. looks good otherwise.

@irtimmer what's your take on enabling this by default sooner or later?

**Comment by irtimmer** [comment]
Latency seems to be biggest problem with encrypted DNS due the extra round trips needed for setting up the encrypted connection. This can still be improved by longer living connections using edns-tcp-keepalive and less round-trips during handshake using TLS 1.3. But this needs also be supported by the servers and TLS 1.3 is not yet supported by GnuTLS. Also there is still an issue if the server doesn't respond with RST when the port is closed, as in that case it will take some time (10s) before the connection is considered unsuccessful and the resolver will fall-back to UDP. This can be solved by checking availability of DNS-over-TLS in the background, but haven't implemented it yet. But after some improvement, when the extra latency is deemed acceptable, it would be nice to have it enabled by default. As beside the extra latency I can't think of any other issues.

**Comment by poettering** [comment]
thanks for the typo fix, let's get this in as soon as the CI likes it.

**Comment by ott** [comment]
@irtimmer We have to develop some general background probing mechanism because we have similar problems with UDP. I will try to write it down properly in the next days and open a separate issue for it. It will likely require some larger changes and some discussion to plan them. I will reference this pull request in the issue.

**Comment by poettering** [comment]
CI all passed. let's merge.

**Comment by poettering** [comment]
See #9277 for a PR with a brief mention of this for the NEWS file for the upcoming v239 release.

File: src/shared/resolve-util.c [hunk 31]
```diff
@@ -25,3 +26,9 @@ static const char* const dnssec_mode_table[_DNSSEC_MODE_MAX] = {
         [DNSSEC_YES] = "yes",
 };
 DEFINE_STRING_TABLE_LOOKUP_WITH_BOOLEAN(dnssec_mode, DnssecMode, DNSSEC_YES);
+
+static const char* const private_dns_mode_table[_PRIVATE_DNS_MODE_MAX] = {
+        [PRIVATE_DNS_NO] = "no",
+        [PRIVATE_DNS_OPPORTUNISTIC] = "opportunistic",
+};
+DEFINE_STRING_TABLE_LOOKUP_WITH_BOOLEAN(private_dns_mode, PrivateDnsMode, PRIVATE_DNS_OPPORTUNISTIC);
```
Comment (yuwata) [inline src/shared/resolve-util.c:34]: So, setting 'yes' is equivalent to 'opportunistic'. Is it intended?

File: src/resolve/resolved-manager.c [hunk 32]
```diff
@@ -1384,6 +1385,15 @@ bool manager_dnssec_supported(Manager *m) {
         return true;
 }
 
+PrivateDnsMode manager_get_private_dns_mode(Manager *m) {
+        assert(m);
+
+        if (m->private_dns_mode != _PRIVATE_DNS_MODE_INVALID)
+                return m->private_dns_mode;
+
+        return _PRIVATE_DNS_MODE_INVALID;
```
Comment (yuwata) [inline src/resolve/resolved-manager.c:1394]: Should this `PRIVATE_DNS_MODE_NO`?

**Comment by keszybz** [comment]
I'm late to the (review) party here. Nevertheless, I think "PrivateDNS=" is a bad name. "DNSTLS=" wasn't pretty, but at least it was informative. "private dns" has the problem that when another mechanism comes along (for example DNS encrypted with dtls or whatever), we'll have a naming problem. "private dns" does not work particularly well in google searches also. Expressing "intent" with low-level settings can be problematic. In particular I don't think this setting should enable e.g. dns-over-dtls if we implement it ever. Hence I think it should remain a low-level setting that describes what it does (DNSTLS= or DNSoverTLS=), and maybe we can add a high-level setting over that later.

**Comment by poettering** [comment]
@keszybz I think I agree. The fact that this PR used the better known and better understood DNS-over-TLS name in the title tells a storry. So yupp DNSOverTLS= or DNSoverTLS= sounds great to me.

**Comment by HLFH** [comment]
@poettering @keszybz I would also advocate for `DNSoverTLS=`.

**Comment by keszybz** [comment]
I think we have agreement to rename this to DNSoverTLS= or DNSOverTLS. "O" is more consistent with other names we have, but "o" is more symmetrical. Not sure what will look better in the end.

@irtimmer Sorry for yanking you back and forth. Any chance you could prepare a patch to do the rename?

**Comment by mbiebl** [comment]
Would it be possible to make this DNS-over-TLS feature compile time optional?
Atm, this pulls in a dependency on GnuTLS into the main systemd package in Debian

**Comment by yuwata** [comment]
I guess `meson -Dgnutls=false` disables DNS-over-TLS.

**Comment by mbiebl** [comment]
We do have ```-Dgnutls=true``` as we want gnutls support in systemd-journal-remote (which is splitt off into a separate package in Debian).

**Comment by mbiebl** [comment]
Even if I build with ```-Ddefault-private-dns=no```, if gnutls has been detected, the DNS-over-TLS code becomes active.

**Comment by keszybz** [comment]
Please open a new issue and mark it for v239.

**Comment by evverx** [comment]
@irtimmer is there any chance you could take a look at https://github.com/systemd/systemd/issues/9511?
